### PR TITLE
feat(ekf_localizer): short periodic updates of z values based on pitch 

### PR DESCRIPTION
**Pull request by SaltUhey**
## Description

Now, Autoware localizer is based on 3DoF(x,y,yaw) EKF estimation at 50 Hz, NDT scan matching at 10 Hz and GNSS.
In case of tunnel, only NDT scan matching is responsible for updating z, roll, and pitch.
Due to it, I faced to localization fault.
Details is Here(https://github.com/autowarefoundation/autoware.universe/issues/4209).

**Conventional method**
conventional_method
**Proposed method**
proposed_method
## Related links
https://github.com/autowarefoundation/autoware.universe/issues/4209

## Tests performed

1. Ruuning Autoware in the environments with slopes
2. Confirm z value improves

## Notes for reviewers
If you need the data with slopes, I gave you.
But data is confidential, so I can give it for TIER IV member only.

## Interface changes
None

## Effects on system behavior
None

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Review by kminoda — COMMENTED** [review]
Thank you for the PR!

File: localization/ekf_localizer/src/ekf_localizer.cpp [hunk 1]
```diff
@@ -343,7 +347,11 @@ void EKFLocalizer::callbackPoseWithCovariance(
   if (!is_activated_) {
     return;
   }
-
+  /* Considering change of z value due to NDT delay*/
+  const rclcpp::Time t_curr = this->now();
+  const double delay_time = (t_curr - msg->header.stamp).seconds();
+  const double dz_delay = considering_z_ndt_delay(current_ekf_twist_, delay_time);
+  msg->pose.pose.position.z += dz_delay;
```
Comment (kminoda) [inline localization/ekf_localizer/src/ekf_localizer.cpp:354]: Please change the implementation here. I'd prefer not to update msg.z with compensated z, since it is kind of a "lie" information.

File: localization/ekf_localizer/include/ekf_localizer/ekf_localizer.hpp [hunk 2]
```diff
@@ -87,6 +87,19 @@ class Simple1DFilter
...
+    double dz = val_sin * vx * t;
+    x_ = x_ + dz;
+    return;
+  };
+  void update_pitch_add(double pitch_rate, double t)
+  {
+    double dp = pitch_rate * t;
+    x_ = x_ + dp;
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (SaltUhey) [reply localization/ekf_localizer/include/ekf_localizer/ekf_localizer.hpp]: OK. I understand.
Based on your suggestion, I would like to make the following changes.
```c++
pitch_filter_.update_pitch_add(pitch_rate, dt);
```
to
```c++
 const double current_pitch = pitch_filter_.get_x();
 const double new_pitch = current_pitch + pitch_rate_ * dt * cntTimerCallback_;
 cntTimerCallback_++;
```
Maybe, `update()` is optimized for updating when ndt is received, so I think it should be separated from `update()` to update with a short EKF cycle.

Each variable is listed below.
cntTimerCallback_: The number of times timercallback() is run until the value of ndt is received
dt: EKF period

File: localization/ekf_localizer/src/ekf_localizer.cpp (hunk 1 above)
Comment (SaltUhey) [reply localization/ekf_localizer/src/ekf_localizer.cpp]: OK.
I will try to implement it in one of the following two functions.
- `measurementUpdatePose()`
- `updateSimple1DFilters()`

File: localization/ekf_localizer/include/ekf_localizer/ekf_localizer.hpp [hunk 3]
```diff
@@ -154,6 +154,9 @@ class EKFLocalizer : public rclcpp::Node
 
   double ekf_rate_;
   double ekf_dt_;
+  double pitch_from_ndt_;
+  double pitch_rate_;
+  int cntTimerCallback_;
```
Comment (kminoda) [inline localization/ekf_localizer/include/ekf_localizer/ekf_localizer.hpp:159]: Please use snake_case for variables, not camelCase.

File: localization/ekf_localizer/src/ekf_localizer.cpp [hunk 4]
```diff
@@ -632,7 +648,17 @@ void EKFLocalizer::initSimple1DFilters(const geometry_msgs::msg::PoseWithCovaria
...
+  const geometry_msgs::msg::TwistStamped & twist, const double delay_time)
+{
+  const double vx = twist.twist.linear.x;
+  const double pitch_rad = pitch_from_ndt_;
+  const double val_sin = std::sin(-pitch_rad);
 
+  const double dz = val_sin * vx * delay_time;
+
+  return dz;
+}
```
Comment (kminoda) [inline localization/ekf_localizer/src/ekf_localizer.cpp:661]: How about something like `calculateDeltaZFromPitch` ?
The current function name sounds a bit different from what it actually does :+1:

File: localization/ekf_localizer/src/ekf_localizer.cpp [hunk 5]
```diff
@@ -466,6 +478,8 @@ void EKFLocalizer::measurementUpdateTwist(
       "twist frame_id must be base_link");
   }
 
+  pitch_rate_ = twist.twist.twist.angular.y;
```
Comment (kminoda) [inline localization/ekf_localizer/src/ekf_localizer.cpp:481]: Why is this here in the first place? :thinking: 

Why not just extracting `twist.twist.twist.angular.y` from `twist_queue_` in e.g. line 207? 

Generally speaking, it is better to avoid using member variables, and this pitch_rate_ member variable seems not necessary.

File: localization/ekf_localizer/src/ekf_localizer.cpp [hunk 6]
```diff
@@ -207,26 +207,33 @@ void EKFLocalizer::timerCallback()
...
   const double roll = roll_filter_.get_x();
-  const double pitch = pitch_filter_.get_x();
   const double yaw = biased_yaw + yaw_bias;
   const double vx = ekf_.getXelement(IDX::VX);
   const double wz = ekf_.getXelement(IDX::WZ);
 
+  const double z = z_filter_.get_x();
+  const double val_sin = -std::sin(pitch_from_ndt_);
+  const double z_addition = val_sin * vx * dt * cntTimerCallback_;
+  const double new_z = z + z_addition;
```
Comment (kminoda) [inline localization/ekf_localizer/src/ekf_localizer.cpp:226]: It is a bit difficult to catch up what is going on in this field. (Not only your fault, but mainly due to the current EKF's implementation, sorry for that :pray: )

This is just an idea, but what about this?
```cpp
  /* pose measurement update */
  if (!pose_queue_.empty()) {
    ...
  }

  /* twist measurement update */
  if (!twist_queue_.empty()) {
    ...
  }

  /* z and pitch update for slopes */
  // UPDATE Z AND PITCH HERE

  /* Extract the results */
  const double x = ekf_.getXelement(IDX::X);
  ... (rest of the code)
```

Comment (SaltUhey) [reply localization/ekf_localizer/src/ekf_localizer.cpp]: It's a good idea.
I will implement your advice.

File: localization/ekf_localizer/src/ekf_localizer.cpp (hunk 5 above)
Comment (SaltUhey) [reply localization/ekf_localizer/src/ekf_localizer.cpp:487]: OK.
But I don't know how to extract `twist.twist.twist.angular.y` from t`wist_queue_` in below code unit.
```c++
    // save the initial size because the queue size can change in the loop
    const size_t n = twist_queue_.size();
    for (size_t i = 0; i < n; ++i) {
      const auto twist = twist_queue_.pop_increment_age();
      measurementUpdateTwist(*twist);
    }
    DEBUG_INFO(get_logger(), "[EKF] measurementUpdateTwist calc time = %f [ms]", stop_watch_.toc());
    DEBUG_INFO(get_logger(), "------------------------- end Twist -------------------------\n");
  }
``` 
For example, `twist_queue_[0].twist.twist.angular.y`?

File: localization/ekf_localizer/include/ekf_localizer/ekf_localizer.hpp [hunk 7]
```diff
@@ -154,6 +154,9 @@ class EKFLocalizer : public rclcpp::Node
 
   double ekf_rate_;
   double ekf_dt_;
+  double pitch_from_ndt_;
+  double pitch_rate_;
+  rclcpp::Time t_receive_pose;
```
Comment (kminoda) [inline localization/ekf_localizer/include/ekf_localizer/ekf_localizer.hpp:159]: Always add underscore after the member variables, it may confuse other developers.
```suggestion
  rclcpp::Time t_receive_pose_;
```

**Comment by YamatoAndo** [comment]
@SaltUhey Could you please add an explanation about this functionality to [the README of the ekf_localizer](https://github.com/autowarefoundation/autoware.universe/blob/main/localization/ekf_localizer/README.md)?

File: localization/ekf_localizer/include/ekf_localizer/ekf_localizer.hpp (hunk 7 above)
Comment (SaltUhey) [reply localization/ekf_localizer/include/ekf_localizer/ekf_localizer.hpp]: > @SaltUhey Could you please add an explanation about this functionality to [the README of the ekf_localizer](https://github.com/autowarefoundation/autoware.universe/blob/main/localization/ekf_localizer/README.md)?

OK. I will add an explanation about this functionality to [the README of the ekf_localizer](https://github.com/autowarefoundation/autoware.universe/blob/main/localization/ekf_localizer/README.md).
Do I provide an additional explanation in [Functions](https://github.com/autowarefoundation/autoware.universe/blob/main/localization/ekf_localizer/README.md#functions)?

**Comment by SaltUhey** [comment]
> @SaltUhey Could you please add an explanation about this functionality to [the README of the ekf_localizer](https://github.com/autowarefoundation/autoware.universe/blob/main/localization/ekf_localizer/README.md)?

OK. I will add an explanation about this functionality to [the README of the ekf_localizer](https://github.com/autowarefoundation/autoware.universe/blob/main/localization/ekf_localizer/README.md).
Do I provide an additional explanation in [Functions](https://github.com/autowarefoundation/autoware.universe/blob/main/localization/ekf_localizer/README.md#functions)?

File: localization/ekf_localizer/src/ekf_localizer.cpp [hunk 8]
```diff
@@ -446,7 +446,15 @@ void EKFLocalizer::measurementUpdatePose(const geometry_msgs::msg::PoseWithCovar
     poseMeasurementCovariance(pose.pose.covariance, params_.pose_smoothing_steps);
 
   ekf_.updateWithDelay(y, C, R, delay_step);
-  updateSimple1DFilters(pose, params_.pose_smoothing_steps);
+
+  // Considering change of z value due to measurement pose delay
+  const auto rpy = tier4_autoware_utils::getRPY(pose.pose.pose.orientation);
+  const double dz_delay = calculateDeltaZFromPitch(current_ekf_twist_, delay_time, rpy.y);
```
Comment (kminoda) [inline localization/ekf_localizer/src/ekf_localizer.cpp:452]: IMHO you don't need to create a dedicated function for this, since the internal calculation can be written in one liner.

File: localization/ekf_localizer/README.md [hunk 9]
```diff
@@ -18,6 +18,7 @@ This package includes the following features:
 - **Automatic estimation of yaw bias** prevents modeling errors caused by sensor mounting angle errors, which can improve estimation accuracy.
 - **Mahalanobis distance gate** enables probabilistic outlier detection to determine which inputs should be used or ignored.
 - **Smooth update**, the Kalman Filter measurement update is typically performed when a measurement is obtained, but it can cause large changes in the estimated value, especially for low-frequency measurements. Since the algorithm can consider the measurement time, the measurement data can be divided into multiple pieces and integrated smoothly while maintaining consistency (see the following figure).
+- **Calculation of vertical correction amount from pitch** mitigates localization instability on slopes. For example, when going uphill, it behaves as if it is buried in the ground as shown in the left side of the figure because EKF only considers 3DoF(x,y,yaw). Therefore, correct the z-coordinate according to the formula shown on the right side of the figure (see the following figure).
```
Comment (kminoda) [inline localization/ekf_localizer/README.md:21]:
```suggestion
- **Calculation of vertical correction amount from pitch** mitigates localization instability on slopes. For example, when going uphill, it behaves as if it is buried in the ground (see the left side of the "Calculate delta from pitch" figure) because EKF only considers 3DoF(x,y,yaw). Therefore, EKF corrects the z-coordinate according to the formula (see the right side of the "Calculate delta from pitch" figure) .
```

**Review by kminoda — APPROVED** [review]
LGTM!

- Confirmed that all the comments have been addressed
- Confirmed that the z interpolation actually mitigates the jumpy estimation of original EKF with our data ([TIER IV INTERNAL LINK](https://tier4.atlassian.net/wiki/spaces/~727523079/pages/2864120230/2023+08+22+EKF+slope+patch))

**Review by YamatoAndo — APPROVED** [review]
